src/server/analytic_services/prometheus_reporting.js

```diff
@@ -2,359 +2,423 @@
-    },
-    generate_default_set: true,
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'bucket_healthy',
-    configuration: {
-        name: get_metric_name('bucket_status'),
-        help: 'Bucket Health',
-        labelNames: ['bucket_name'],
-    },
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'bucket_capacity_precent',
-    configuration: {
-        name: get_metric_name('bucket_capacity'),
-        help: 'Bucket Capacity Precent',
-        labelNames: ['bucket_name'],
-    },
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'bucket_quota_precent',
-    configuration: {
-        name: get_metric_name('bucket_quota'),
-        help: 'Bucket Quota Precent',
-        labelNames: ['bucket_name'],
-    },
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'resource_healthy',
-    configuration: {
-        name: get_metric_name('resource_status'),
-        help: 'Resource Health',
-        labelNames: ['resource_name'],
-    },
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'system_links',
-    configuration: {
-        name: get_metric_name('system_links'),
-        help: 'System Links',
-        labelNames: ['resources', 'buckets', 'dashboard'],
-    },
-}, {
-    metric_type: 'Gauge',
-    metric_variable: 'health_status',
-    configuration: {
-        name: get_metric_name('health_status'),
-        help: 'Health status',
-    },
-    generate_default_set: true,
-}]);
-
-
-class PrometheusReporting {
-    static instance() {
-        PrometheusReporting._instance = PrometheusReporting._instance || new PrometheusReporting();
-        return PrometheusReporting._instance;
+
+    export_metrics() {
+        return this._registry.metrics();
     }
+}
 
+// -----------------------------------------
+// Report containing node js metrics
+// collected and provided by the prom client
+// -----------------------------------------
+class NodeJsReport extends BasePrometheusReport {
     constructor() {
-        METRIC_RECORDS.filter(metric => metric.generate_default_set).forEach(metric => {
-            this[`set_${metric.metric_variable}`] = function(data) {
-                if (!this.enabled()) return;
-                this._metrics[metric.metric_variable].set(data);
-            };
+        super();
+
+        this._prom_client.collectDefaultMetrics({
+            register: this._registry,
+            prefix: this.metric_prefix
         });
-        this._prom_client = prom_client;
-        if (this.enabled()) {
-            this.define_metrics();
-        }
     }
 
-    client() {
-        return this._prom_client;
+    get metric_prefix() {
+        return `${super.metric_prefix}${dbg.get_process_name()}_`;
     }
+}
 
-    enabled() {
-        return config.PROMETHEUS_ENABLED;
-    }
+register_report(
+    config.NODEJS_METRICS_REPORT,
+    NodeJsReport
+);
 
-    define_metrics() {
-        this._metrics = {};
-        prom_client.collectDefaultMetrics({ prefix: get_metric_name('defaults_') });
```

Comment:
isn't removing this means everything will be spilled directly on NooBaa_ ?
If so, I don't think its a good choice.

Reply:
This is a bug. I missed it in the refactoring. 
Thank you

Reply:
my mistake. this is not a bug each process gets it's own prefix using the process name. So it will be
`NooBaa_WebServer_*` for example

Reply:
How would endpoints look like if there are more than 1 ?
What would happen to customer metrics on webserver ? Path change ?

Reply:
each endpoint export its own metrics, same prefix, diff pod name.
The new path for  custom metrics is `/metrics/web_server/noobaa_core`.

In general, the path template for every metric  served from the `web_srever ` is `/metrics/<procsse_name>/<report_name>`
To consume all  metrics from all reports for a process the path will be `/metrics/<procsse_name>`

But in this PR I still support the old path `/metrics` and `/metrics/counter`

Reply:
I understand the route
I'm wandering about the prometheus side.

How would metrics, 
- lets say NooBaa_X previously would look like now
- metrics from the endpoint in case more than 1 exists

Reply:
Ok, let me explain:
- The only metrics that were consumed are the custom ones from the code which stayed exactly the same just served from a new route.
- the default metrics change their prefix, but nobody that I know of ever consumed them
- The endpoint metrics will be differentiated by pod name/id (all of them will have the `NooBaa_Endpoint_*` prefix

Reply:
Ok, can we have the node.js metrics in the manner of
NooBaa_Endpoint_<ID>_NODE_

and core
NooBaa_NODE_

and bg
NooBaa_BG_NODE_

etc, ?

there's quite a lot of counters on the default node.js part, lets group them under a sub topic

Reply:
What do you mean? we do not have control over the individual default nodejs metric.
They are defined and managed by the prom client

Reply:
Also, what does NODE represent in your answer? I do not get it. The name of the default metric is defined by the client

Reply:
Can you paste, from prometheus the full one (metric name and value) for one of the core custom metrics and one of the core node.js metrics ?

Reply:
Ok, I am pasting one metric from each report that we expose:

NodeJS report for WebServer process:
`NooBaa_WebServer_nodejs_eventloop_lag_seconds 0.006712859`

NodeJS report for BgWorker process :
`NooBaa_BGWorkers_nodejs_eventloop_lag_seconds 0.014447736`

NodeJS report for HostedAgents process:
`NooBaa_HostedAgents_nodejs_eventloop_lag_seconds 7.513233968`

NodeJS report for Endpoint process:
`NooBaa_Endpoint_nodejs_eventloop_lag_seconds 2.703802685`

NooBaa Core report (custom metrics):
`NooBaa_providers_bandwidth_read_size{type="KUBERNETES"} 2347008`
(the prefix here is only "NooBaa" in order to not break current consumers)

NooBaa Endpoint report (custom metrics):
will start with  `NooBaa_` but the full matrix can be decided when we will go to implement it

Reply:
ok, and we plan to keep NooBaa_ for all webserver's custom metrics right ? 
Then all is good if the answer is yes

Reply:
The answer is YES, 
I would have loved to address `core` to the custom metrics but I don't want to break existing consumers